File: dotnet/SketchOxide/src/Sampling/ReservoirSampling.cs
using System;
using System.Text;
using SketchOxide.Native;

namespace SketchOxide.Sampling;

/// <summary>
/// Reservoir sampling for maintaining a uniform random sample of k items from a stream.
///
/// Uses Algorithm R (Vitter) to maintain a reservoir of k items such that at any point,
/// each item in the stream has an equal probability of being in the sample.
/// </summary>
/// <remarks>
/// Reservoir sampling is useful when you need a representative sample from a stream
/// of unknown (or very large) size without storing all items. The algorithm guarantees
/// that each item has exactly k/n probability of being in the sample (where n is
/// the total number of items seen).
/// </remarks>
public sealed class ReservoirSampling : NativeSketch
{
    private readonly uint _k;

    /// <summary>
    /// Creates a new reservoir sampler that maintains k items.
    /// </summary>
    /// <param name="k">The size of the reservoir (number of items to sample). Must be greater than 0.</param>
    /// <exception cref="ArgumentOutOfRangeException">Thrown if k is 0.</exception>
    /// <exception cref="OutOfMemoryException">Thrown if native memory allocation fails.</exception>
    public ReservoirSampling(uint k)
    {
        if (k == 0)
            throw new ArgumentOutOfRangeException(nameof(k), k, "K must be greater than 0");

        _k = k;
        NativePtr = SketchOxideNative.reservoirsampling_new(k);

        if (NativePtr == 0)
            throw new OutOfMemoryException("Failed to allocate native ReservoirSampling");
    }

    /// <summary>
    /// Private constructor for deserialization.
    /// </summary>
    private ReservoirSampling(uint k, nuint ptr)
    {
        _k = k;
        NativePtr = ptr;
    }

    /// <summary>
    /// Gets the reservoir size (k) of this sampler.
    /// </summary>
    public uint K
    {
        get
        {
            CheckAlive();
            return _k;
        }
    }

    /// <summary>
    /// Updates the reservoir with a new item from the stream.
    /// The item may or may not be added to the reservoir based on the sampling algorithm.
    /// </summary>
    /// <param name="data">The bytes representing the item.</param>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sampler is disposed.</exception>
    public void Update(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        unsafe
        {
            fixed (byte* ptr = data)
            {
                SketchOxideNative.reservoirsampling_update(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Updates the reservoir with a string item from the stream.
    /// </summary>
    /// <param name="value">The string item.</param>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sampler is disposed.</exception>
    public void Update(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        Update(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Returns the current sample as an array of byte arrays.
    /// </summary>
    /// <returns>The sampled items. May contain fewer than k items if fewer have been seen.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sampler is disposed.</exception>
    public byte[][] Sample()
    {
        CheckAlive();

        byte[] rawData = SketchOxideNative.reservoirsampling_sample(NativePtr, out uint count);

        if (rawData == null || rawData.Length == 0 || count == 0)
            return Array.Empty<byte[]>();

        var result = new byte[count][];

        // Parse the serialized data format: [len(4 bytes), item(len bytes)]...
        int offset = 0;
        for (uint i = 0; i < count && offset < rawData.Length; i++)
        {
            if (offset + 4 > rawData.Length) break;

            // Read item length (4 bytes, little-endian)
            int itemLen = BitConverter.ToInt32(rawData, offset);
            offset += 4;

            if (offset + itemLen > rawData.Length) break;

            // Read item bytes
            result[i] = new byte[itemLen];
            Array.Copy(rawData, offset, result[i], 0, itemLen);
            offset += itemLen;
        }

        return result;
    }

    /// <summary>
    /// Returns the current sample as an array of strings.
    /// </summary>
    /// <returns>The sampled items decoded as UTF-8 strings.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sampler is disposed.</exception>
    public string[] SampleAsStrings()
    {
        byte[][] bytes = Sample();
        var result = new string[bytes.Length];
        for (int i = 0; i < bytes.Length; i++)
        {
            result[i] = Encoding.UTF8.GetString(bytes[i]);
        }
        return result;
    }

    /// <summary>
    /// Gets the number of items currently in the reservoir.
    /// This may be less than k if fewer items have been seen.
    /// </summary>
    public int Count
    {
        get
        {
            CheckAlive();
            return Sample().Length;
        }
    }

    /// <summary>
    /// Serializes the sampler to a byte array.
    /// </summary>
    /// <returns>Serialized sampler bytes.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sampler is disposed.</exception>
    public byte[] Serialize()
    {
        CheckAlive();
        return SketchOxideNative.reservoirsampling_serialize(NativePtr, out _);
    }

    /// <summary>
    /// Deserializes a reservoir sampler from a byte array.
    /// </summary>
    /// <param name="data">Serialized sampler bytes.</param>
    /// <param name="k">The k parameter used when creating the original sampler.</param>
    /// <returns>A new ReservoirSampling instance.</returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ArgumentException">Thrown if data is invalid.</exception>
    public static ReservoirSampling Deserialize(byte[] data, uint k)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));

        nuint ptr = SketchOxideNative.reservoirsampling_deserialize(data, (ulong)data.Length);
        if (ptr == 0)
            throw new ArgumentException("Failed to deserialize ReservoirSampling: invalid data");

        return new ReservoirSampling(k, ptr);
    }

    /// <summary>
    /// Returns a string representation of the sampler.
    /// </summary>
    public override string ToString()
    {
        if (IsDisposed)
            return "ReservoirSampling(disposed)";
        return $"ReservoirSampling(k={_k}, count={Count})";
    }

    /// <summary>
    /// Frees the native Rust instance.
    /// </summary>
    protected override void FreeNative()
    {
        if (NativePtr != 0)
        {
            SketchOxideNative.reservoirsampling_free(NativePtr);
            NativePtr = 0;
        }
    }
}

File: java/src/main/java/com/sketches_oxide/sampling/ReservoirSampling.java
package com.sketches_oxide.sampling;

import com.sketches_oxide.NativeSketch;
import com.sketches_oxide.native.SketchOxideNative;

/**
 * ReservoirSampling - Uniform Random Sampling from Streams.
 *
 * Maintains a uniform random sample of k items from a stream of unknown length.
 * Each item in the stream has an equal probability of being in the final sample.
 * Classic algorithm by Vitter (1985) for one-pass sampling.
 *
 * <p><strong>Algorithm:</strong>
 * <ul>
 *   <li>First k items go directly into the reservoir</li>
 *   <li>Item i (i &gt; k) replaces a random item with probability k/i</li>
 *   <li>Guarantees uniform sampling without knowing stream size</li>
 *   <li>Single pass, O(k) memory</li>
 * </ul>
 *
 * <p><strong>Properties:</strong>
 * <ul>
 *   <li>Space: O(k) items</li>
 *   <li>Time: O(1) per item</li>
 *   <li>Uniformity: Each item has exactly k/n probability of being sampled</li>
 * </ul>
 *
 * <p><strong>Use cases:</strong>
 * <ul>
 *   <li>Sampling from large data streams</li>
 *   <li>Random subset selection for analysis</li>
 *   <li>A/B testing cohort selection</li>
 *   <li>Statistical sampling for surveys</li>
 *   <li>Downsampling for visualization</li>
 * </ul>
 *
 * <p><strong>Example:</strong>
 * <pre>
 * // Keep a random sample of 100 items
 * try (ReservoirSampling sampler = new ReservoirSampling(100)) {
 *     // Process a stream of items
 *     for (String item : largeStream) {
 *         sampler.update(item.getBytes());
 *     }
 *
 *     // Get the uniform random sample
 *     byte[][] sample = sampler.sample();
 *     System.out.printf("Sampled %d items from %d total%n",
 *         sample.length, sampler.count());
 *
 *     // Process sampled items
 *     for (byte[] item : sample) {
 *         String str = new String(item);
 *         System.out.println("Sampled: " + str);
 *     }
 * }
 * </pre>
 *
 * @see VarOptSampling
 * @see <a href="https://en.wikipedia.org/wiki/Reservoir_sampling">Reservoir Sampling</a>
 */
public final class ReservoirSampling extends NativeSketch {

    private final long k;

    /**
     * Create a new ReservoirSampling instance.
     *
     * @param k the reservoir size (maximum number of samples to keep)
     * @throws IllegalArgumentException if k &lt;= 0
     */
    public ReservoirSampling(long k) {
        if (k <= 0) {
            throw new IllegalArgumentException("k must be positive, got: " + k);
        }

        this.k = k;
        this.nativePtr = SketchOxideNative.reservoirsampling_new(k);

        if (this.nativePtr == 0) {
            throw new OutOfMemoryException("Failed to allocate ReservoirSampling");
        }
    }

    /**
     * Add an item to the stream.
     *
     * The item may or may not be kept in the reservoir based on
     * the random selection algorithm.
     *
     * @param item the item to potentially sample
     * @throws IllegalStateException if the sampler has been closed
     * @throws NullPointerException if item is null
     */
    public void update(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        SketchOxideNative.reservoirsampling_update(nativePtr, item);
    }

    /**
     * Add a string item to the stream.
     *
     * @param item the string item to potentially sample
     */
    public void update(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        update(item.getBytes());
    }

    /**
     * Get the current sample.
     *
     * Returns up to k items uniformly sampled from the stream.
     * If fewer than k items have been seen, returns all of them.
     *
     * @return array of sampled items
     * @throws IllegalStateException if the sampler has been closed
     */
    public byte[][] sample() {
        checkAlive();
        return SketchOxideNative.reservoirsampling_sample(nativePtr);
    }

    /**
     * Get the current sample as strings.
     *
     * Convenience method that converts sampled bytes to strings.
     *
     * @return array of sampled items as strings
     */
    public String[] sampleAsStrings() {
        byte[][] bytes = sample();
        String[] result = new String[bytes.length];
        for (int i = 0; i < bytes.length; i++) {
            result[i] = new String(bytes[i]);
        }
        return result;
    }

    /**
     * Get the number of items currently in the reservoir.
     *
     * @return number of items in sample (at most k)
     */
    public int len() {
        byte[][] s = sample();
        return s.length;
    }

    /**
     * Get the reservoir capacity.
     *
     * @return the k parameter
     */
    public long capacity() {
        return k;
    }

    /**
     * Get the total count of items seen.
     *
     * @return total number of items added to the stream
     * @throws IllegalStateException if the sampler has been closed
     */
    public long count() {
        checkAlive();
        return SketchOxideNative.reservoirsampling_count(nativePtr);
    }

    /**
     * Check if the reservoir is full.
     *
     * @return true if at least k items have been seen
     */
    public boolean isFull() {
        return count() >= k;
    }

    /**
     * Get the probability that any given item is in the sample.
     *
     * @return k/n where n is the number of items seen
     */
    public double getSamplingProbability() {
        long n = count();
        if (n == 0) return 0.0;
        return Math.min(1.0, (double) k / n);
    }

    /**
     * Serialize the sampler to binary format.
     *
     * @return binary representation of the sampler
     * @throws IllegalStateException if the sampler has been closed
     */
    public byte[] serialize() {
        checkAlive();
        return SketchOxideNative.reservoirsampling_serialize(nativePtr);
    }

    /**
     * Deserialize from binary format.
     *
     * @param data the serialized data from {@link #serialize()}
     * @return a new ReservoirSampling instance
     * @throws IllegalArgumentException if data is invalid
     * @throws NullPointerException if data is null
     */
    public static ReservoirSampling deserialize(byte[] data) {
        if (data == null) {
            throw new NullPointerException("data cannot be null");
        }

        long ptr = SketchOxideNative.reservoirsampling_deserialize(data);
        if (ptr == 0) {
            throw new IllegalArgumentException("Invalid or corrupt serialized ReservoirSampling data");
        }

        // Create with dummy value, then replace pointer
        ReservoirSampling sampler = new ReservoirSampling(100);
        SketchOxideNative.reservoirsampling_free(sampler.nativePtr);
        sampler.nativePtr = ptr;
        return sampler;
    }

    @Override
    public void close() {
        if (nativePtr != 0) {
            freeNative();
            nativePtr = 0;
        }
    }

    @Override
    protected void freeNative() {
        if (nativePtr != 0) {
            SketchOxideNative.reservoirsampling_free(nativePtr);
        }
    }

    @Override
    public String toString() {
        try {
            if (nativePtr != 0) {
                return String.format("ReservoirSampling(k=%d, count=%d, sampleSize=%d)",
                        k, count(), len());
            }
        } catch (IllegalStateException e) {
            // Closed
        }
        return "ReservoirSampling(closed)";
    }
}
